[C10D] Report detected failures when emitting collective end events.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109739
* #111239
* #111246

We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

kumpera committed:
[C10D] Report detected failures when emitting collective end events.

We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

wconstab commented on test/distributed/test_hooks.py:
```diff
@@ -138,6 +141,65 @@ def check_op(idx, coll_name):
+        dist.all_gather(tensor_list, tensor)
+
+        tensor2 = torch.ones([2, 3]).cuda(self.rank) * self.rank
+        if self.rank == 2:
```
are we assuming world_size == 2? is there a testing macro we can wrap to ensure that if so?

kumpera replied:
the test harness hardcode world size 4

wconstab commented on test/distributed/test_hooks.py:
```diff
@@ -138,6 +141,65 @@ def check_op(idx, coll_name):
+        def check_op(idx, coll_name, end_fail=False):
+            self.assertEqual(default_pg_name, starts[idx].pg_name)
+            self.assertEqual(self.backend_name, starts[idx].backend)
+            self.assertGreaterEqual(starts[idx].sequence_number, 0)
```
why not assertEqual(sequence_number, idx)?

wconstab commented on test/distributed/test_hooks.py:
```diff
@@ -138,6 +141,65 @@ def check_op(idx, coll_name):
+            self.assertEqual(coll_name, starts[idx].operation)
+            self.assertEqual(starts[idx].error_message, None)
+            if end_fail:
+                self.assertIsNotNone(ends[idx].error_message, None)
```
perhaps assert some known string is present in the error msg too?
(and that might also answer my question: what error are we expecting in this example?)

kumpera replied:
good point, hopefully it will be reliable enough during testing.

kumpera committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

kumpera commented:
There's one major issue left on this PR: it doesn't work with gloo.

kumpera committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

fduwjj reviewed (commented):
How we ensure WatchDog does not abort the program and let us return error message?

wconstab committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

wconstab committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/Backend.cpp:
```diff
@@ -18,6 +29,9 @@ void commonEventinit(
   evt.operation = c10d::opTypeToString(work.retrieveOpType());
   evt.drop_count = 0;
+  // isCompleted is mutable :facepalm:
+  if (const_cast<Work&>(work).isCompleted() && !work.isSuccess())
```
@fduwjj I think we should discuss this part

we're calling work.exception() which will query some cuda APIs.

we're doing this from the main watchdog thread.

i think that in today's watchdog design this is already something we're doing in the main loop, so i don't have a big reservation about it.

but if you're redesigning the loops to put cuda accesses on another thread, you'd also want to move this part.  would be good to think through how we'll handle the callback reporting error message safely, or else maybe not add the error to the callback in the first place.  (I think it should be possible to do though)

wconstab committed:
Update on "[C10D] Report detected failures when emitting collective end events."


We leverage the fact that errors get tracked in Work to report them.

The error message by itself is not crazy useful and we might need to
include some form of error code in the mix (nccl failure, timeout, etc).

[ghstack-poisoned]
